cpp/oneapi/dal/backend/primitives/rng/device_engine_dpc.cpp

```diff
@@ -0,0 +1,155 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/rng/device_engine.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+
+#include <oneapi/mkl.hpp>
+
+namespace oneapi::dal::backend::primitives {
+
+namespace bk = oneapi::dal::backend;
+
+template <typename Distribution, typename Type>
+sycl::event generate_rng(Distribution& distr,
+                         device_engine& engine_,
+                         std::int64_t count,
+                         Type* dst,
+                         const event_vector& deps) {
+    switch (engine_.get_device_engine_base_ptr()->get_engine_type()) {
+        case engine_type::mt2203: {
+            auto& device_engine =
+                *(dynamic_cast<gen_mt2203*>(engine_.get_device_engine_base_ptr().get()))->get();
+            return oneapi::mkl::rng::generate(distr, device_engine, count, dst, deps);
+        }
+        case engine_type::mcg59: {
+            auto& device_engine =
+                *(dynamic_cast<gen_mcg59*>(engine_.get_device_engine_base_ptr().get()))->get();
+            return oneapi::mkl::rng::generate(distr, device_engine, count, dst, deps);
+        }
+        case engine_type::mrg32k3a: {
+            auto& device_engine =
+                *(dynamic_cast<gen_mrg32k*>(engine_.get_device_engine_base_ptr().get()))->get();
+            return oneapi::mkl::rng::generate(distr, device_engine, count, dst, deps);
+        }
+        case engine_type::philox4x32x10: {
+            auto& device_engine =
+                *(dynamic_cast<gen_philox*>(engine_.get_device_engine_base_ptr().get()))->get();
+            return oneapi::mkl::rng::generate(distr, device_engine, count, dst, deps);
+        }
+        case engine_type::mt19937: {
+            auto& device_engine =
+                *(dynamic_cast<gen_mt19937*>(engine_.get_device_engine_base_ptr().get()))->get();
+            return oneapi::mkl::rng::generate(distr, device_engine, count, dst, deps);
+        }
+        default: throw std::runtime_error("Unsupported engine type in generate_rng");
+    }
+}
+
+template <typename Type>
+void uniform(sycl::queue& queue,
+             std::int64_t count,
+             Type* dst,
+             device_engine& engine_,
+             Type a,
+             Type b,
+             const event_vector& deps) {
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) == sycl::usm::alloc::host) {
+        throw domain_error(dal::detail::error_messages::unsupported_data_type());
+    }
+    oneapi::mkl::rng::uniform<Type> distr(a, b);
+    auto event = generate_rng(distr, engine_, count, dst, deps);
+    event.wait_and_throw();
+    engine_.skip_ahead_cpu(count);
+}
+
+//Currently only CPU impl
```

Comment:
Please add more meaningful comments.
What is the `buffer` is not obvious, for example.
```suggestion
/// Generates a sequence of uniformly distributed random numbers from the interval [a, b).
/// The numbers in the sequence do not repeat.
/// In current version the data generation is only available on CPU.
///
/// @tparam Type ...
///
/// @param queue ...
```